### PR TITLE
[stable/velero] Upgrade Ark 0.10.x to Velero 0.11

### PR DESCRIPTION
**Pull request by josephgorse**
#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by josephgorse** [comment]
@mattfarina

**Comment by josephgorse** [comment]
@scottrigby @unguiculus @domcar

**Comment by nrb** [comment]
Hi @josephgorse, thanks for tackling this! I'm on the Velero team and would like to help out here. I'm putting this PR on my list to review tomorrow.

**Comment by unguiculus** [comment]
Please note that the old chart at `stable/ark` must be deprecated and should not be removed.

See https://github.com/helm/charts/blob/master/PROCESSES.md#promoting-a-chart-from-incubator-to-stable

The same rules apply.

**Comment by josephgorse** [comment]
@nrb @unguiculus Fixed the Ark deprecation issue. Can someone approve to test?

File: stable/velero/README.md
Comment (nrb) [inline stable/velero/README.md]: This should be v0.11.0, and instructions for upgrading from v0.10.x to v0.11.0 can be found at https://heptio.github.io/velero/v0.11.0/migrating-to-velero

Comment (nrb) [reply stable/velero/README.md]: Doh, I see those are further down below.

IMO, the v0.10 section instructions should be removed.

Comment (josephgorse) [reply stable/velero/README.md]: Done.

File: stable/velero/templates/backupstoragelocation.yaml
Comment (nrb) [inline stable/velero/templates/backupstoragelocation.yaml]: This should be de-dented, see PR #11430.

File: stable/velero/templates/restic-daemonset.yaml
Comment (nrb) [inline stable/velero/templates/restic-daemonset.yaml:19]: Add the following from PR #11430 

```
 {{- with .Values.podAnnotations }}
      annotations:
{{ toYaml . | indent 8 }}
      {{- end }}```

File: stable/velero/templates/restic-daemonset.yaml
Comment (nrb) [inline stable/velero/templates/restic-daemonset.yaml:91]: Add the following from PR 11430

```
          {{- with .Values.resources }}
          resources:
{{ toYaml . | indent 12 }}
          {{- end }}
```

**Review by nrb — COMMENTED** [review]
A couple things I noticed from another PR that can be incorporated here.

Also, this is something that I observed when trying to create all objects for a Velero install with a single YAML file, but I think the chart may still be susceptible to timing issues from #11637.

I don't have a whole lot of experience with Helm, but I'm wondering if there's a way to address this within this PR.

**Comment by nrb** [comment]
> I don't have a whole lot of experience with Helm, but I'm wondering if there's a way to address this within this PR.

I see you have the `crd-install` hook, which addressed the problem. Thanks!

File: stable/velero/templates/backupstoragelocation.yaml
Comment (josephgorse) [reply stable/velero/templates/backupstoragelocation.yaml]: Done.

File: stable/velero/templates/restic-daemonset.yaml
Comment (josephgorse) [reply stable/velero/templates/restic-daemonset.yaml:19]: Done

File: stable/velero/templates/restic-daemonset.yaml
Comment (josephgorse) [reply stable/velero/templates/restic-daemonset.yaml:91]: Done

**Comment by davidq2q** [comment]
Please merge this one. Heptio ark chart not working propperly right now.

File: stable/velero/values.yaml
Comment (PabloCastellano) [inline stable/velero/values.yaml]: The `prefix` parameter should be below backupStorageLocation (not below config). Check https://github.com/helm/charts/pull/11430 which also fixes this

File: stable/velero/templates/serverstatusrequests.yaml
Comment (PabloCastellano) [inline stable/velero/templates/serverstatusrequests.yaml]: Missing endline

File: stable/velero/values.yaml
Comment (josephgorse) [reply stable/velero/values.yaml]: Done

File: stable/velero/templates/serverstatusrequests.yaml
Comment (josephgorse) [reply stable/velero/templates/serverstatusrequests.yaml]: Done

**Comment by unguiculus** [comment]
Since this is a major update anyways, I suggest to also apply the new Kubernetes label conventions. See https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md#names-and-labels

**Comment by josephgorse** [comment]
@unguiculus @nrb I'm not that familiar with the history of the chart (don't actually work for Heptio :upside_down_face:), but I think the latest should conform to the label guidelines while keeping the intended functionality intact.

**Comment by unguiculus** [comment]
I added a few thing if you don't mind. Please verify.

File: stable/velero/templates/deployment.yaml
Comment (nrb) [inline stable/velero/templates/deployment.yaml]: This is inconsistent with Velero's upstream versions.

@skriss what do you think about getting these RBAC and deployment versions onto v1 for 1.0?

Comment (skriss) [reply stable/velero/templates/deployment.yaml]: yeah makes sense, users can always tweak manually if they're running on old versions.

File: stable/ark/Chart.yaml
Comment (unguiculus) [inline stable/ark/Chart.yaml]: We should just bump the patch version here.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/ark/Chart.yaml
Comment (josephgorse) [reply stable/ark/Chart.yaml]: Its a completely new chart, shouldn't we start again at 1.0.0 for velero and leave ark at 4.2.1?

Comment (josephgorse) [reply stable/ark/Chart.yaml]: Done

File: stable/velero/values.yaml
Comment (nrb) [inline stable/velero/values.yaml]: Update to v0.11.0.

File: stable/velero/templates/secret.yaml
Comment (nrb) [inline stable/velero/templates/secret.yaml:13]: I had trouble getting this format to work, though I may have been doing something wrong...

Here's what worked for me, and the command I used to do the install:

```
{{- if and .Values.credentials.useSecret (not .Values.credentials.existingSecret) -}}
apiVersion: v1
kind: Secret
metadata:
  name: {{ include "velero.fullname" . }}
  labels:
    app.kubernetes.io/name: {{ include "velero.name" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    helm.sh/chart: {{ include "velero.chart" . }}
type: Opaque
data:
  cloud: |-
    {{ .Values.credentials.secretContents | b64enc }}
{{- end }}
```

`helm install --name velero -f values.yaml --set-file credentials.secretContents=/home/nrb/credentials-velero-gke ./velero`

File: stable/velero/templates/NOTES.txt
Comment (nrb) [inline stable/velero/templates/NOTES.txt:1]: I would add the follow here so that users can actually verify.

```
Check that the velero is up and running:

    kubectl get deployment/{{ .Release.Name }} -n {{ .Release.Namespace }}

Check that the secret has been created:

    kubectl get secret/{{ include "velero.fullname" . }} -n {{ .Release.Namespace }}
```

File: stable/velero/templates/secret.yaml
Comment (nrb) [inline stable/velero/templates/secret.yaml:13]: I wasn't able to get this format to work. I instead used the following, along with the associated command:

```
{{- if and .Values.credentials.useSecret (not .Values.credentials.existingSecret) -}}
apiVersion: v1
kind: Secret
metadata:
  name: {{ include "velero.fullname" . }}
  labels:
    app.kubernetes.io/name: {{ include "velero.name" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    helm.sh/chart: {{ include "velero.chart" . }}
type: Opaque
data:
  cloud: |-
    {{ .Values.credentials.secretContents | b64enc }}
{{- end }}
```
`helm install --name velero --namespace velero -f values.yaml --set-file credentials.secretContents=/home/nrb/credentials-velero-gke ./velero`

File: stable/velero/values.yaml
Comment (nrb) [inline stable/velero/values.yaml]: Let's update this URL to v0.11.0.

**Review by nrb — COMMENTED** [review]
A couple more notes as I've had more time to play with this.

I may just be misunderstanding how the secret creation works, so if someone has an example that works with the existing template I'd appreciate it.

File: stable/velero/templates/deployment.yaml
Comment (ctrox) [inline stable/velero/templates/deployment.yaml]: Noticed this while testing, if you enable metrics, it will template `true` into the annotations. I guess this should be `{{- with .Values.metrics.podAnnotations }}`?

File: stable/velero/templates/secret.yaml
Comment (josephgorse) [reply stable/velero/templates/secret.yaml:13]: My only concern here is that `cloud: |` is now hard-coded in. Is there any case where a secret would not start with that? I took the secrets template exactly from the old Ark chart, so it should work?

I've used the secret template as it is in the PR now (`{{- range $key, $value := .Values.credentials.secretContents }}`) and didn't have any issues with it. My secret in values.yaml looked something like this
```
  secretContents:
    cloud: |
      [default]
      aws_access_key_id = ABC123
      aws_secret_access_key = ABC123
```

File: stable/velero/templates/deployment.yaml
Comment (josephgorse) [reply stable/velero/templates/deployment.yaml]: Done

File: stable/velero/values.yaml
Comment (josephgorse) [reply stable/velero/values.yaml]: Done

File: stable/velero/templates/NOTES.txt
Comment (josephgorse) [reply stable/velero/templates/NOTES.txt:1]: Done

File: stable/velero/templates/secret.yaml
Comment (nrb) [reply stable/velero/templates/secret.yaml:13]: Grr, somehow github duplicated some review comments. Sorry!

Comment (nrb) [reply stable/velero/templates/secret.yaml:13]: That's true that it would be hard coded...all of our documentation includes it, but I don't think it's mandatory.


I was unable to get my GCP JSON working in my own values.yaml with the current code (and hadn't in the old Ark template), but I believe it may have been an indenting issue. I'll look more today as it's likely better to keep the current `range` functionality, as you say, and I'm formatting my values incorrectly.

**Comment by josephgorse** [comment]
@unguiculus It looks like the e2e test failure is because of a dirty test environment? CRDs already exist in the environment before the chart being tested is installed. Can you help with this?

**Comment by nrb** [comment]
We cut a `v1.0.0-alpha1` today, so this chart is still lagging a bit. I think we'll want the v0.11.x chart, but we'll be following up with a chart for 1.0, too.

**Comment by ekaley** [comment]
@nrb Lets merge this, its been an open PR for 49 days, we should get this merged and cut over to velero (0.11.0 came out FEB 28!!!...its April 17). All the little fixes we can merge in in smaller PR's.  This is too big of a change to have on a PR branch for this long.

**Comment by nrb** [comment]
@ekaley I am not a maintainer on this repository.

**Comment by josephgorse** [comment]
@nrb @unguiculus If I'm not mistaken, the PR needs approval from someone from Heptio and someone from Helm? What can we do to get this thing merged?

**Review by nrb — APPROVED** [review]
> @nrb @unguiculus If I'm not mistaken, the PR needs approval from someone from Heptio and someone from Helm? What can we do to get this thing merged?

I don't believe so - this hasn't been the case in the past. That said, we would like to be more involved and potentially take ownership of the chart going forward.

I've marked my approval, though I'm coming up to speed on Helm and am not familiar with the test failures.

**Comment by unguiculus** [comment]
I had to clean up CRDs on the test cluster.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**
